pkg/core/object/metadata.go

```diff
@@ -197,3 +200,641 @@ func VerifyHeaderForMetadata(hdr object.Object) error {
+		}
+		dbVal = []byte(hex.EncodeToString(dbVal))
+	case object.FilterSplitID:
+		if len(dbVal) != 16 {
+			return nil, nil, fmt.Errorf("invalid split ID len %d != 16", len(dbVal))
+		}
+		uid, err := uuid.Parse(fltVal)
+		if err == nil {
+			return dbVal, uid[:], nil
+		}
+		copy(uid[:], dbVal)
+		dbVal = []byte(uid.String())
+	}
+	return dbVal, []byte(fltVal), nil
+}
+
+// matches object attribute's search query value to the DB-stored one. Matcher
+// must be supported but not [object.MatchNotPresent] or numeric.
+func matchValues(dbVal []byte, matcher object.SearchMatchType, fltVal []byte) bool {
+	switch {
+	default:
+		return false // TODO: check whether supported in blindlyProcess. Then panic here
+	case matcher == object.MatchNotPresent || IsIntegerSearchOp(matcher):
+		panic(fmt.Sprintf("unexpected matcher %s", matcher))
+	case matcher == object.MatchStringEqual:
+		return bytes.Equal(dbVal, fltVal)
+	case matcher == object.MatchStringNotEqual:
+		return !bytes.Equal(dbVal, fltVal)
+	case matcher == object.MatchCommonPrefix:
+		return bytes.HasPrefix(dbVal, fltVal)
+	}
+}
+
+func restoreIntAttribute(b []byte) (string, error) {
+	if len(b) != intValLen {
+		return "", fmt.Errorf("invalid len %d", len(b))
+	}
+	switch b[0] {
+	default:
+		return "", fmt.Errorf("invalid sign byte %d", b[0])
+	case 1:
+		return new(big.Int).SetBytes(b[1:]).String(), nil
+	case 0:
+		cp := slices.Clone(b[1:])
+		for i := range cp {
+			cp[i] = ^cp[i]
+		}
+		n := new(big.Int).SetBytes(cp)
+		return n.Neg(n).String(), nil
+	}
+}
+
+func restoreAttributeValue(attr string, stored []byte) (string, error) {
+	switch attr {
+	case object.FilterOwnerID, object.FilterFirstSplitObject, object.FilterParentID:
+		return base58.Encode(stored), nil
+	case object.FilterPayloadChecksum, object.FilterPayloadHomomorphicHash:
+		return hex.EncodeToString(stored), nil
+	case object.FilterSplitID:
+		uid, err := uuid.ParseBytes(stored)
+		if err != nil {
+			return "", invalidMetaBucketKeyErr([]byte{metaPrefixAttrIDPlain}, fmt.Errorf("decode split ID: decode UUID: %w", err))
+		}
+		return uid.String(), nil
+	}
+	return string(stored), nil
+}
+
+func intMatches(dbVal *big.Int, matcher object.SearchMatchType, fltVal *big.Int) bool {
+	switch matcher {
+	default:
+		panic(fmt.Errorf("unexpected integer matcher %d", matcher))
+	case object.MatchNumGT:
+		return dbVal.Cmp(fltVal) > 0
+	case object.MatchNumGE:
+		return dbVal.Cmp(fltVal) >= 0
+	case object.MatchNumLT:
+		return dbVal.Cmp(fltVal) < 0
+	case object.MatchNumLE:
+		return dbVal.Cmp(fltVal) <= 0
+	}
+}
+
+// PreprocessSearchQuery accepts verified search filters, requested attributes
+// along and continuation cursor, verifies the cursor and returns additional
+// arguments to pass into [DB.Search]. If the query is valid but unreachable,
+// [ErrUnreachableQuery] is returned.
+func PreprocessSearchQuery(fs object.SearchFilters, attrs []string, cursor string) (*SearchCursor, map[int]ParsedIntFilter, error) {
```

Comment:
This `map[int]ParsedIntFilter` is also not very helpful, it can be like `object.SearchFilters -> InternalSearchFilters` with this data embedded into appropriate fields.

Reply:
https://github.com/nspcc-dev/neofs-node/blob/092ff01ee240397eda24d76105bb9c89760742f6/pkg/local_object_storage/metabase/metadata.go#L217

Reply:
I know, I know, but you're touching it and it's a part of `MetaDataKVHandler` complexity problem as well.